led_strip/src/led_strip_spi_dev.c

```diff
@@ -0,0 +1,213 @@
+    *(buf + 0) |= data & BIT(7) ? BIT(7) | BIT(6) : BIT(7);
+}
+
+static esp_err_t led_strip_spi_set_pixel(led_strip_t *strip, uint32_t index, uint32_t red, uint32_t green, uint32_t blue)
+{
+    led_strip_spi_obj *spi_strip = __containerof(strip, led_strip_spi_obj, base);
+    ESP_RETURN_ON_FALSE(index < spi_strip->strip_len, ESP_ERR_INVALID_ARG, TAG, "index out of maximum number of LEDs");
+    uint32_t start = index * spi_strip->bytes_per_pixel * LED_COLOR_BUF;
+    memset(spi_strip->pixel_buf + start, 0, spi_strip->bytes_per_pixel * LED_COLOR_BUF);
+    __led_strip_spi_bit(green, &spi_strip->pixel_buf[start]);
+    __led_strip_spi_bit(red, &spi_strip->pixel_buf[start + LED_COLOR_BUF]);
+    __led_strip_spi_bit(blue, &spi_strip->pixel_buf[start + LED_COLOR_BUF * 2]);
+
+    return ESP_OK;
+}
+
+static esp_err_t led_strip_spi_set_pixel_rgbw(led_strip_t *strip, uint32_t index, uint32_t red, uint32_t green, uint32_t blue, uint32_t white)
+{
+    led_strip_spi_obj *spi_strip = __containerof(strip, led_strip_spi_obj, base);
+    ESP_RETURN_ON_FALSE(index < spi_strip->strip_len, ESP_ERR_INVALID_ARG, TAG, "index out of maximum number of LEDs");
+    ESP_RETURN_ON_FALSE(spi_strip->bytes_per_pixel == 4, ESP_ERR_INVALID_ARG, TAG, "wrong LED pixel format, expected 4 bytes per pixel");
+    uint32_t start = index * spi_strip->bytes_per_pixel * LED_COLOR_BUF;
+    // SK6812 component order is GRBW
+    memset(spi_strip->pixel_buf + start, 0, spi_strip->bytes_per_pixel * LED_COLOR_BUF);
+    __led_strip_spi_bit(green, &spi_strip->pixel_buf[start]);
+    __led_strip_spi_bit(red, &spi_strip->pixel_buf[start + LED_COLOR_BUF]);
+    __led_strip_spi_bit(blue, &spi_strip->pixel_buf[start + LED_COLOR_BUF * 2]);
+    __led_strip_spi_bit(blue, &spi_strip->pixel_buf[start + LED_COLOR_BUF * 3]);
+
+    return ESP_OK;
+}
+
+static esp_err_t led_strip_spi_refresh(led_strip_t *strip)
+{
+    led_strip_spi_obj *spi_strip = __containerof(strip, led_strip_spi_obj, base);
+    spi_transaction_t tx_conf;
+    memset(&tx_conf, 0, sizeof(tx_conf));
+
+    tx_conf.length = spi_strip->strip_len * spi_strip->bytes_per_pixel * LED_BUF_BIT;
+    tx_conf.tx_buffer = spi_strip->pixel_buf;
+    tx_conf.rx_buffer = NULL;
+    ESP_RETURN_ON_ERROR(spi_device_transmit(spi_strip->spi_device, &tx_conf), TAG, "transmit pixels by SPI failed");
+
+    return ESP_OK;
+}
+
+static esp_err_t led_strip_spi_clear(led_strip_t *strip)
+{
+    led_strip_spi_obj *spi_strip = __containerof(strip, led_strip_spi_obj, base);
+    // Write zero to turn off all leds
+    if (spi_strip->bytes_per_pixel == 3) {
+        for (int index = 0; index < spi_strip->strip_len; index++) {
+            led_strip_spi_set_pixel(strip, index, 0, 0, 0);
+        }
+    } else if (spi_strip->bytes_per_pixel == 4) {
+        for (int index = 0; index < spi_strip->strip_len; index++) {
+            led_strip_spi_set_pixel_rgbw(strip, index, 0, 0, 0, 0);
+        }
+    }
```

Comment:
I think this can be cimplified by calling  `__led_strip_spi_bit`  for each byte in the `pixel_buf`?